File: src/chapters/chapter-05/csharp/Models/UserProfile.cs
using System.ComponentModel;

using Microsoft.Extensions.AI;
using Microsoft.Extensions.VectorData;

using Microsoft.SemanticKernel.Data;

namespace AdvancedAIShoppingAssistant.Models;

public class UserProfile
{
    [Description("User id of the user.")]
    public string? UserId { get; set; }

    [Description("Username of the user.")]
    public string? Name { get; set; }

    [Description("Email address of the user.")]
    public string? Email { get; set; }

    [Description("Brand affinity of the user.")]
    public string? BrandAffinity { get; set; }

    [Description("Budget limit of the user.")]
    public decimal? Budget { get; set; }

    [Description("Category interests of the user.")]
    public List<string> CategoryInterests { get; set; } = new();

    [Description("Latest visited products by the user.")]
    public List<string> LatestVisitedProducts { get; set; } = new();
}

public class UserProfileVectorStore
{
    [VectorStoreKey]

    public ulong UserPofileId { get; set; }

    [VectorStoreData]
    [TextSearchResultLink]

    public string? UserId { get; set; }

    [VectorStoreData]
    [TextSearchResultName]
    public string? UserName { get; set; }

    [VectorStoreData]
    [TextSearchResultValue]
    public string? UserProfileAsText { get; set; }

    [VectorStoreVector(Dimensions: 4, DistanceFunction = DistanceFunction.CosineSimilarity, IndexKind = IndexKind.Hnsw)]
    public Embedding<float>? UserProfileAsEmbedding { get; set; }
}

File: src/chapters/chapter-05/python/models/user_profile.py
from semantic_kernel.functions import kernel_function
from typing import List, Optional
from dataclasses import dataclass, field

@dataclass
class UserProfile:

    user_id: str = field(metadata={"description":"User ID of the user."})
    name: str = field(metadata={"description":"Username of the user."})
    email: str = field(metadata={"description":"Email address of the user."})
    brand_affinity: str = field(metadata={"description": "Brand affinity of the user."})
    budget: float = field(metadata={"description": "Budget limit of the user."})
    category_interests: List[str] = field(default_factory=list,metadata={"description":"User's interested categories"})
    latest_visited_products: List[str] = field(default_factory=list,metadata={"description":"Recently visited products"})